File: go/prune.go
package main

import (
	"regexp"
	"strings"

	gpuprofiling "github.com/iteratorlee/samprof-client/go/go-gen"
)

type CriticalNodeType int

const (
	CRITICAL_TYPE_PY_BACKWARD CriticalNodeType = iota
	CRITICAL_TYPE_PY_LOSS
	CRITICAL_TYPE_PY_FORWARD
	CRITICAL_TYPE_TORCH_OP
	CRITICAL_TYPE_TF_OP
	CRITICAL_TYPE_LEAF
	NOT_CRITICAL_NODE
)

func isCriticalNode(node *gpuprofiling.CPUCallingContextNode) CriticalNodeType {
	nodeName := node.GetFuncName()
	if !strings.Contains(nodeName, "python3") {
		if strings.Contains(nodeName, "backward") {
			// fmt.Printf("%v : %v\n", nodeName, "CRITICAL_TYPE_PY_BACKWARD")
			return CRITICAL_TYPE_PY_BACKWARD
		}
		if strings.Contains(nodeName, "loss") {
			// fmt.Printf("%v : %v\n", nodeName, "CRITICAL_TYPE_PY_LOSS")
			//fixme haven't considered pyFileName
			return CRITICAL_TYPE_PY_LOSS
		}
		if strings.Contains(nodeName, "forward") {
			// fmt.Printf("%v : %v\n", nodeName, "CRITICAL_TYPE_PY_FORWARD")
			return CRITICAL_TYPE_PY_FORWARD
		}
	}

	torchOPRegex, _ := regexp.Compile("at::_ops::(\\S+)::call(\\S+)")
	tfOPRegex, _ := regexp.Compile("(\\S+)Op(Kernel)?.+::Compute")
	if match := torchOPRegex.MatchString(nodeName); match {
		// fmt.Printf("%v : %v\n", nodeName, "CRITICAL_TYPE_TORCH_OP")
		return CRITICAL_TYPE_TORCH_OP
	}

	if match := tfOPRegex.MatchString(nodeName); match {
		// fmt.Printf("%v : %v\n", nodeName, "CRITICAL_TYPE_TF_OP")
		return CRITICAL_TYPE_TF_OP
	}

	// fmt.Printf("%v : %v\n", nodeName, "NOT_CRITICAL_NODE")
	return NOT_CRITICAL_NODE
}

func pruneCCT(cpuCCT *gpuprofiling.CPUCallingContextTree) {
	rootId := int64(cpuCCT.GetRootID())
	id2node := cpuCCT.GetNodeMap()

	id2ChildIDs := make(map[int64][]uint64)
	// set parentID
	for nid, node := range id2node {
		// root Node has no parent
		if nid != int64(rootId) {
			for true {
				parentId := int64(node.GetParentID())
				// always keep root node
				if parentId == rootId {
					break
				}
				if parentNode := id2node[parentId]; isCriticalNode(parentNode) != NOT_CRITICAL_NODE {
					break
				} else {
					node.ParentID = parentNode.GetParentID()
				}
			}
		}
		if isCriticalNode(node) != NOT_CRITICAL_NODE {
			if childs, ok := id2ChildIDs[int64(node.GetParentID())]; ok {
				childs = append(childs, uint64(nid))
			} else {
				id2ChildIDs[int64(node.GetParentID())] = []uint64{uint64(nid)}
			}
		}
	}
	for nid, node := range id2node {
		node.ChildIDs = id2ChildIDs[nid]
	}
}

File: py/prune.py
import logging
import python_gen as gpuprofiling
import re
from enum import Enum
class CriticalNodeType(Enum):
    CRITICAL_TYPE_PY_BACKWARD=1
    CRITICAL_TYPE_PY_LOSS=2
    CRITICAL_TYPE_PY_FORWARD=3
    CRITICAL_TYPE_TORCH_OP=4
    CRITICAL_TYPE_TF_OP=5
    CRITICAL_TYPE_LEAF=6
    NOT_CRITICAL_NODE=7

def isCriticalNode(node : gpuprofiling.CPUCallingContextNode)->CriticalNodeType:
    nodeName = node.funcName
    if "python3" not in nodeName:
        if 'backward' in nodeName:
            logging.debug(f'{nodeName} is a py backward node')
            return CriticalNodeType.CRITICAL_TYPE_PY_BACKWARD
        elif 'loss' in nodeName:
            logging.debug(f'{nodeName} is a py loss node')
            return CriticalNodeType.CRITICAL_TYPE_PY_LOSS
        elif 'forward' in nodeName:
            logging.debug(f'{nodeName} is a py forward node')
            return CriticalNodeType.CRITICAL_TYPE_PY_FORWARD

    torchOPRegex ="at::_ops::(\\S+)::call(\\S+)"
    tfOPRegex ="(\\S+)Op(Kernel)?.+::Compute"
    if re.match(torchOPRegex, nodeName):
        logging.debug(f'{nodeName} is a torch op')
        return CriticalNodeType.CRITICAL_TYPE_TORCH_OP
    elif re.match(tfOPRegex, nodeName):
        logging.debug(f'{nodeName} is a tf op')
        return CriticalNodeType.CRITICAL_TYPE_TF_OP

    if len(node.childs) == 0:
        logging.debug(f'{nodeName} is a leaf node')
        return CriticalNodeType.CRITICAL_TYPE_LEAF

    logging.debug(f'{nodeName} is not a critical node')
    return CriticalNodeType.NOT_CRITICAL_NODE

def pruneCCT(cpuCCT :gpuprofiling.CPUCallingContextTree):
    rootId = cpuCCT.rootID
    id2node = cpuCCT.nodeMap
    id2childIDs={}
    #skip all non critical nodes
    for nid,node in id2node.items():
        if nid != rootId:
            while True:
                parentId = node.parentID
                if parentId == rootId:
                    break
                parentnode = id2node[parentId]
                if isCriticalNode(parentnode) != CriticalNodeType.NOT_CRITICAL_NODE:
                    break
                else :
                    node.parentID = parentnode.parentID
        if isCriticalNode(node)!=CriticalNodeType.NOT_CRITICAL_NODE:
            parentId = node.parentID
            if parentId not in id2childIDs:
                id2childIDs[parentId]=[]
            id2childIDs[parentId].append(nid)
    #only keep the critical nodes
    for nid,node in id2node.items():
        while len(node.childIDs)>0:
            node.childIDs.pop()
        if nid in id2childIDs:
            for childId in id2childIDs[nid]:
                node.childIDs.append(childId)
